File: wear/src/main/java/ro/upt/sma/heart/data/SensorDataStore.java
package ro.upt.sma.heart.data;

import static android.content.Context.SENSOR_SERVICE;

import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import java.util.HashMap;

public class SensorDataStore {

  public interface HeartRateListener {
    void onValueChanged(int value);
  }

  private final SensorManager sensorManager;
  private final Sensor heartRateSensor;
  private final HashMap<HeartRateListener, SensorEventListener> listeners = new HashMap<>();

  public SensorDataStore(Context context) {
    this.sensorManager = ((SensorManager) context.getSystemService(SENSOR_SERVICE));
    this.heartRateSensor = sensorManager.getDefaultSensor(Sensor.TYPE_HEART_RATE);
  }

  public void registerHeartRateListener(final HeartRateListener listener) {
    SensorEventListener eventListener = new SensorEventListener() {
      @Override
      public void onSensorChanged(SensorEvent sensorEvent) {
        listener.onValueChanged((int) sensorEvent.values[0]);
      }

      @Override
      public void onAccuracyChanged(Sensor sensor, int i) {
      }
    };

    if (heartRateSensor != null) {
      sensorManager.registerListener(eventListener, heartRateSensor, SensorManager.SENSOR_DELAY_FASTEST);
      listeners.put(listener, eventListener);
    }
  }

  public void unregisterHeartRateListener(HeartRateListener listener) {
    if (listeners.containsKey(listener)) {
      SensorEventListener eventListener = listeners.get(listener);
      sensorManager.unregisterListener(eventListener);
      listeners.remove(listener);
    }
  }

}

File: wear/src/main/java/ro/upt/sma/heart/data/SensorDataStore.kt
package ro.upt.sma.heart.data

import android.content.Context.SENSOR_SERVICE

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import java.util.HashMap

class SensorDataStore(context: Context) {

    private val sensorManager: SensorManager = context.getSystemService(SENSOR_SERVICE) as SensorManager
    private val heartRateSensor = sensorManager.getDefaultSensor(Sensor.TYPE_HEART_RATE)

    private val listeners = HashMap<HeartRateListener, SensorEventListener>()

    interface HeartRateListener {
        fun onValueChanged(value: Int)
    }

    fun registerHeartRateListener(listener: HeartRateListener) {
        val eventListener = object : SensorEventListener {
            override fun onSensorChanged(sensorEvent: SensorEvent) {
                listener.onValueChanged(sensorEvent.values[0].toInt())
            }

            override fun onAccuracyChanged(sensor: Sensor, i: Int) {}
        }

        if (heartRateSensor != null) {
            sensorManager.registerListener(eventListener, heartRateSensor, SensorManager.SENSOR_DELAY_FASTEST)
            listeners[listener] = eventListener
        }
    }

    fun unregisterHeartRateListener(listener: HeartRateListener) {
        if (listeners.containsKey(listener)) {
            val eventListener = listeners[listener]
            sensorManager.unregisterListener(eventListener)
            listeners.remove(listener)
        }
    }

}
